### PR TITLE
JBDS-3849 use promises to fetch requirements

### PR DESCRIPTION
**Pull request by nickboldt**
JBDS-3849 use promise to fetch requirements into cache folder and copy from cache to dist

### CODE REVIEW

File: gulpfile.js [hunk 1]
```diff
@@ -286,22 +308,37 @@ gulp.task('prefetch', function(cb) {
...
-          console.log('DOWNLOADED <- ' + currentFile);
+          counter++;
+          promises[counter] = copyFromPrefetchToDist(currentFile, destFile);
+          counter--;
         }
       });
+
     }
   }
+  promise.all(promises);
```
Comment (dgolovin) [inline gulpfile.js]: Nick it should be something like this snipped below and coounter would not be neded anymore 
``` JavaScript
promise.all(promises).then((result)=> {
  cb();
}).catch((error)=>{
  cb(error);
});
```

File: requirements.json [hunk 2]
```diff
@@ -12,7 +12,7 @@
   "oc.zip": {
```
Comment (dgolovin) [inline requirements.json]: Please send separate PR for this change.

**Comment by nickboldt** [comment]
SHA fix for requirements moved to #159 at Denis' request.

**Comment by nickboldt** [comment]
So, this PR works if you run `gulp prefetch && gulp dist`

But if you start from a clean checkout (no prefetch folder) or do `gulp clean-all && gulp dist` the sequence looks like this:

http://pastebin.com/7jg3merG

File: gulpfile.js [hunk 3]
```diff
@@ -269,14 +270,34 @@ function resolveInstallerExePath(artifactType) {
...
+    });
+  }
+  return aPromise;
+}
+
+// prefetch all the installer dependencies so we can package them up into the .exe
 gulp.task('prefetch', function(cb) {
-  let counter=0;
+  let promises = [];
+  if (!fs.existsSync(prefetchFolder)) { mkdirp(prefetchFolder, cb); }
```
Comment (dgolovin) [inline gulpfile.js]: This is the reason why, you call back here and the in promise.all().then().catch();

**Comment by nickboldt** [comment]
With this last suggestion, build works.

Here's a console log (expires in 1 week):

http://pastebin.com/WpAjCp2U

**Comment by dgolovin** [comment]
Fix changes a lot and not easy to follow, instead is should be simple promise wrap around download we already have. Here is an example https://github.com/redhat-developer-tooling/developer-platform-install/compare/master...dgolovin:JBDS-3849?expand=1.
I didn't test it really and it could look much better after refactoring, but I think that is Pavol's idea in rough implementation. It just get rid of ugly counter and introduce Promise.all() call.

File: gulpfile.js [hunk 4]
```diff
@@ -17,6 +17,7 @@ var gulp = require('gulp'),
     minimatch = require('minimatch'),
     copy = require('gulp-copy'),
     concat = require('gulp-concat'),
+    promise = require('promise'),
```
Comment (ppitonak) [inline gulpfile.js]: Promises are natively supported by Node.js 4 and 5 so you don't need this dependency.

File: gulpfile.js [hunk 5]
```diff
@@ -269,14 +270,34 @@ function resolveInstallerExePath(artifactType) {
   return path.join(zaRoot, artifactName + '-' + pjson.version + artifactType + '-installer.exe');
 }
 
-// download all the installer dependencies so we can package them up into the .exe
+function copyFromPrefetchToDist(currentFile, destFile) {
```
Comment (ppitonak) [inline gulpfile.js]: I see no reason why we need to copy all downloaded bits. We create 7z archive like this `7za.exe a installer.7z installer-dir/*`, can't we just add second path to this command? I would be both faster and would save 1GB on hard disk

`7za.exe a installer.7z installer-dir/* prefetch-dependencies/*`

File: gulpfile.js [hunk 6]
```diff
@@ -79,16 +80,16 @@ gulp.task('create-dist-win-dir', function(cb) {
...
+  var cmd = path.join('node_modules', '.bin') + path.sep + 'electron-packager . ' + artifactName + ' --platform=' + artifactPlatform + ' --arch=' + artifactArch;
   cmd += ' --version=' + electronVersion + ' --out="' + buildFolderPath + '" --overwrite --asar=true';
   cmd += ' --version-string.CompanyName="Red Hat, Inc."';
   cmd += ' --version-string.ProductName="' + pjson.productName + '"';
   cmd += ' --version-string.OriginalFilename="' + artifactName + '-' + pjson.version + '-installer.exe"';
   cmd += ' --version-string.FileDescription="' + pjson.description + ' v' + pjson.version + '"';
   cmd += ' --app-copyright="Copyright 2016 Red Hat, Inc."';
   cmd += ' --app-version="' + pjson.version + '"' + ' --build-version="' + pjson.version + '"';
-  cmd += ' --prune --ignore=test';
+  cmd += ' --prune --ignore="test|prefetch-dependencies"';
```
Comment (dgolovin) [inline gulpfile.js]: This seems is not working for me, so I got all the prefetch-dependencies inside first light installer..

File: gulpfile.js [hunk 7]
```diff
@@ -285,23 +306,33 @@ gulp.task('prefetch', function(cb) {
...
+
     }
   }
+  promise.all(promises).then((result)=> {
+    cb();
+  }).catch((error)=>{
+    cb(error);
+  });
+
   installerExe = resolveInstallerExePath('-bundle');
```
Comment (ppitonak) [inline gulpfile.js]: This should go inside then statement.

File: gulpfile.js [hunk 8]
```diff
@@ -285,23 +306,33 @@ gulp.task('prefetch', function(cb) {
...
           }
         } else {
-          console.log('DOWNLOADED <- ' + currentFile);
+          promises.push(copyFromPrefetchToDist(currentFile, destFile));
         }
       });
+
     }
   }
+  promise.all(promises).then((result)=> {
```
Comment (ppitonak) [inline gulpfile.js]: You need to put `return` statement in front of promise.all so that Gulp handles sequence of tasks correctly.

File: package.json [hunk 9]
```diff
@@ -64,8 +64,10 @@
     "karma-junit-reporter": "0.4.1",
     "karma-ng-html2js-preprocessor": "0.2.1",
     "minimatch": "3.0.0",
+    "mkdirp": "^0.5.1",
     "mocha": "2.4.5",
     "mock-fs": "3.8.0",
+    "promise": "^7.1.1",
```
Comment (ppitonak) [inline package.json]: not necessary (see comment above)

File: package.json [hunk 10]
```diff
@@ -64,8 +64,10 @@
     "karma-junit-reporter": "0.4.1",
     "karma-ng-html2js-preprocessor": "0.2.1",
     "minimatch": "3.0.0",
+    "mkdirp": "^0.5.1",
```
Comment (ppitonak) [inline package.json]: Not really necessary with Node.js 5.x but has to be listed explicitly with Node.js 4.x. I created a PR https://github.com/redhat-developer-tooling/developer-platform-install/pull/160 but we can add it with this PR, it doesn't matter.

**Comment by ppitonak** [comment]
I added comments directly to Nick's code. I still need to try it

File: gulpfile.js [hunk 11]
```diff
@@ -215,25 +216,25 @@ function createSHA256File(filename, cb) {
...
+  runSequence(['check-requirements', 'clean'], 'create-dist-win-dir', ['generate',
     'prepare-tools'], 'package', 'cleanup', cb);
-});
+});
 
   // Create bundled installer
 gulp.task('package-bundle', function(cb) {
-  runSequence(['check-requirements', 'clean'], 'create-dist-win-dir', ['generate',
+  runSequence(['check-requirements', 'clean'], 'create-dist-win-dir', ['generate',
     'prepare-tools'], 'prefetch', 'package', 'cleanup', cb);
```
Comment (ppitonak) [inline gulpfile.js]: we should be able to run _prefetch_ in parallel with _generate_ and _prepare-tools_, it will speed up the build a little bit

File: gulpfile.js [hunk 12]
```diff
@@ -215,25 +216,25 @@ function createSHA256File(filename, cb) {
...
-});
+});
 
 // Create both installers
 gulp.task('dist', function(cb) {
-  runSequence(['check-requirements', 'clean'], 'create-dist-win-dir', ['generate',
+  runSequence(['check-requirements', 'clean'], 'create-dist-win-dir', ['generate',
     'prepare-tools'], 'package', 'prefetch', 'package', 'cleanup', cb);
```
Comment (ppitonak) [inline gulpfile.js]: we should be able to run _prefetch_ in parallel with _generate_ and _prepare-tools_, it will speed up the build a little bit

File: gulpfile.js [hunk 13]
```diff
@@ -285,23 +306,33 @@ gulp.task('prefetch', function(cb) {
...
-              });
+            promises.push(new Promise(function(resolve, reject) {
+              console.log('[INFO] Download ' + currentUrl + ' to ' + currentFile);
+              request(currentUrl)
+                .pipe(fs.createWriteStream(currentFile)).on('finish',function(err, res) {
+                  if (err) { 
+                    reject(err);
+                  } else {
+                    // console.log('DOWNLOADED <- ' + currentFile);
+                    promises.push(copyFromPrefetchToDist(currentFile, destFile));
```
Comment (ppitonak) [inline gulpfile.js]: I think that this is causing trouble... you can't push copy promise to _promises_ array because it will be called in parallel with download task. You need to refactor it out and in the end do something like `return Promise.all(downloadFiles).then(copyFiles).then(() => cb()).catch(cb);`

File: gulpfile.js (hunk 6 above)
Comment (ppitonak) [inline gulpfile.js]: Can you try adding parentheses? `--ignore="(test|prefetch-dependencies)"`

File: gulpfile.js (hunk 7 above)
Comment (nickboldt) [inline gulpfile.js]: You mean before the cb() on line 331 ?

File: gulpfile.js (hunk 11 above)
Comment (dgolovin) [inline gulpfile.js]: that is doable and should work good, because download is going to be done often for jbds bits only, probably

File: gulpfile.js (hunk 5 above)
Comment (dgolovin) [inline gulpfile.js]: good point, it should work

File: gulpfile.js (hunk 7 above)
Comment (ppitonak) [inline gulpfile.js]: yes

File: gulpfile.js (hunk 11 above)
Comment (ppitonak) [inline gulpfile.js]: I would leave it as is for now and improve performance in another PR

File: package.json (hunk 9 above)
Comment (nickboldt) [inline package.json]: I disagree. I get "ReferenceError: promise is not defined" if I don't declare it in gulpfile.js.

File: gulpfile.js (hunk 4 above)
Comment (nickboldt) [inline gulpfile.js]: I disagree. I get "ReferenceError: promise is not defined" if I don't declare it in gulpfile.js.

Comment (ppitonak) [inline gulpfile.js]: you need to use Promise (with upper-case "P")

File: gulpfile.js [hunk 14]
```diff
@@ -269,40 +275,49 @@ function resolveInstallerExePath(artifactType) {
...
-      });
+        });
+      }));
     }
   }
-  installerExe = resolveInstallerExePath('-bundle');
+  Promise.all(promises).then((result) => {
+    installerExe = resolveInstallerExePath('-bundle');
+    cb();
+  }).catch((error)=>{
```
Comment (dgolovin) [inline gulpfile.js]: should be just 
``` JavaScript
}).catch(cb);
```

File: package.json (hunk 9 above)
Comment (dgolovin) [inline package.json]: promise => Promice should fix the issue.

File: gulpfile.js [hunk 15]
```diff
@@ -269,40 +275,49 @@ function resolveInstallerExePath(artifactType) {
...
-        } else {
-          console.log('DOWNLOADED <- ' + currentFile);
-        }
-      });
+        });
+      }));
     }
   }
-  installerExe = resolveInstallerExePath('-bundle');
+  Promise.all(promises).then((result) => {
```
Comment (dgolovin) [inline gulpfile.js]: you should add return 
``` JavaScript
 return Promise.all(promises).then((result) => {
```

**Comment by dgolovin** [comment]
https://github.com/redhat-developer-tooling/developer-platform-install/pull/161 reworked version of this PR

**Comment by dgolovin** [comment]
See https://github.com/redhat-developer-tooling/developer-platform-install/pull/161 for details.
